python/lsst/ctrl/mpexec/separablePipelineExecutor.py

```diff
@@ -0,0 +1,264 @@
+
+        Parameters
+        ----------
+        graph : `lsst.pipe.base.QuantumGraph`
+            The quantum graph defining the pipeline and datasets to
+            be initialized.
+        register_dataset_types : `bool`, optional
+            If `True`, register all output dataset types from the pipeline
+            represented by ``graph``.
+        save_init_outputs : `bool`, optional
+            If `True`, create init-output datasets in this object's output run.
+        save_versions : `bool`, optional
+            If `True`, save a package versions dataset.
+        """
+        pre_exec_init = PreExecInit(self._butler, self._task_factory, extendRun=self._clobber_output)
+        pre_exec_init.initialize(
+            graph=graph,
+            saveInitOutputs=save_init_outputs,
+            registerDatasetTypes=register_dataset_types,
+            saveVersions=save_versions,
+        )
+
+    def make_pipeline(self, pipeline_uri: Union[str, lsst.resources.ResourcePath]) -> lsst.pipe.base.Pipeline:
+        """Build a pipeline from pipeline and configuration information.
+
+        Parameters
+        ----------
+        pipeline_uri : `str` or `lsst.resources.ResourcePath`
+            URI to a file containing a pipeline definition. A URI fragment may
+            be used to specify a subset of the pipeline, as described in
+            :ref:`pipeline-running-intro`.
+
+        Returns
+        -------
+        pipeline : `lsst.pipe.base.Pipeline`
+            The fully-built pipeline.
+        """
+        return lsst.pipe.base.Pipeline.from_uri(pipeline_uri)
+
+    def make_graph(
+        self, pipeline: lsst.pipe.base.Pipeline, where: str = "", builder: Optional[_GraphBuilderLike] = None
+    ) -> lsst.pipe.base.QuantumGraph:
+        """Build a quantum graph from a pipeline and input datasets.
+
+        Parameters
+        ----------
+        pipeline : `lsst.pipe.base.Pipeline`
+            The pipeline for which to generate a quantum graph.
+        where : `str`, optional
+            A data ID query that constrains the quanta generated.
+        builder : `lsst.pipe.base.GraphBuilder`-like, optional
+            A graph builder that implements a
+            `~lsst.pipe.base.GraphBuilder.makeGraph` method. By default, a new
+            instance of `lsst.pipe.base.GraphBuilder` is used.
+
+        Returns
+        -------
+        graph : `lsst.pipe.base.QuantumGraph`
+            The quantum graph for ``pipeline`` as run on the datasets
+            identified by ``where``.
+
+        Notes
+        -----
+        This method does no special handling of empty quantum graphs. If
+        needed, clients can use `len` to test if the returned graph is empty.
+        """
+        if not builder:
+            builder = lsst.pipe.base.GraphBuilder(
+                self._butler.registry,
+                skipExistingIn=self._skip_existing_in,
+                clobberOutputs=self._clobber_output,
+            )
```

Comment:
Right now I think it's pretty arbitrary what's a constructor argument to `GraphBuilder` vs. a `GraphBuilder.makeGraph` argument.  If you've got any opinions on how we might want to move them around, I'd be happy to take that into account the next time `GraphBuilder` gets reworked or replaced.

That also means I'm not really treating it like an interface that is likely to have other implementations in the future, but I don't see any harm in having this hook anyway - it might turn out to be useful as-is, and I think it'll be pretty easy to modify slightly in the future if that turns out to be more useful.

Reply:
Well, yes, that's the problem with duck-typing: no matter how the class is defined/specified, the client still needs to make assumptions about how its methods are called.

In any case, I don't think I'm in a position to comment on the design. I don't even understand what some of the parameters are for.